### PR TITLE
Thread process groups through training checkpoint paths

### PR DESCRIPTION
**Pull request by yashaswikarnati**
## What
Split from WIP integration PR #5377. This PR carries only the Megatron process-group threading slice needed before the larger hetero MIMO entrypoint work:

- Thread explicit process groups through training reductions, schedule selection, memory logging, and checkpoint save/load metadata.
- Let checkpointing use model-local tp/pp/dp_cp groups when a model exposes a `pg_collection`, while preserving MPU fallback behavior for stock runs.
- Store module-local TP groups on RADIO and SSM modules so sharded checkpointing does not fall back to global tensor-parallel groups.
- Namespace MIMO optimizer sharded keys to avoid per-module distributed-optimizer key collisions.

Intentionally excluded from this split:

- Example entrypoint/bootstrap/data changes.
- RNG key-prefix namespacing from #5377.
- Broader custom setup hooks and throughput-normalization changes from the WIP branch.

## Validation
- `git diff --cached --check`
- `uv run isort megatron/core/models/mimo/optimizer.py megatron/training/utils/common_utils.py`
- `uv run isort --check-only megatron/core/models/mimo/optimizer.py megatron/training/utils/common_utils.py`
- `python3 -m py_compile megatron/core/models/mimo/optimizer.py megatron/core/models/vision/radio.py megatron/core/ssm/gated_delta_net.py megatron/core/ssm/mamba_layer.py megatron/core/ssm/mamba_mixer.py megatron/training/checkpointing.py megatron/training/initialize.py megatron/training/training.py megatron/training/utils/common_utils.py`

### CODE REVIEW

File: megatron/core/models/vision/radio.py [hunk 1]
```diff
@@ -211,6 +211,7 @@ def __init__(
         self.ln_pre = None
         self.ln_post = None
         self.pg_collection = pg_collection
+        self.tp_group = pg_collection.tp if pg_collection is not None else None
```
Comment (yashaswikarnati) [inline megatron/core/models/vision/radio.py]: fallback should be parallel state global tp group

File: megatron/training/utils/common_utils.py [hunk 2]
```diff
@@ -295,8 +294,12 @@ def logical_and_across_model_parallel_group(
     return bool(input.item())
 
 
-def report_memory(name):
-    """Simple GPU memory report."""
+def report_memory(name, group=None):
```
Comment (yashaswikarnati) [inline megatron/training/utils/common_utils.py]: instead of group, can we call this process group

File: megatron/training/checkpointing.py [hunk 3]
```diff
@@ -368,8 +368,18 @@ def read_metadata(tracker_filename):
     return max_iter, release
 
 
-def get_rng_state(ckpt_format: str, tp_group: torch.distributed.ProcessGroup, pp_group: torch.distributed.ProcessGroup) -> Union[List[Dict[str, Any]], ShardedObject]:
-    """Collect rng state across data parallel ranks."""
+def get_rng_state(
+    ckpt_format: str,
+    tp_group: torch.distributed.ProcessGroup,
+    pp_group: torch.distributed.ProcessGroup,
+    key_prefix: str = '',
```
Comment (yashaswikarnati) [inline megatron/training/checkpointing.py]: key prefix should be last arg ?

File: megatron/training/checkpointing.py [hunk 4]
```diff
@@ -378,28 +388,30 @@ def get_rng_state(ckpt_format: str, tp_group: torch.distributed.ProcessGroup, pp
         'cuda_rng_state': torch.cuda.get_rng_state(),
         'rng_tracker_states': tensor_parallel.get_cuda_rng_tracker().get_states()}
 
+    dp_world_size = get_pg_size(dp_cp_group) if dp_cp_group is not None else mpu.get_data_parallel_world_size()
```
Comment (yashaswikarnati) [inline megatron/training/checkpointing.py]: this seems wrong ? this is dp group not dp cp ?

File: megatron/training/training.py [hunk 5]
```diff
@@ -3221,6 +3251,22 @@ def train(
 
             args.no_load_optim = no_load_optim
 
+    # Resolve the language-model process groups when a hetero MultiModule collection
```
Comment (yashaswikarnati) [inline megatron/training/training.py]: remove comment # Resolve the language-model process groups when a hetero MultiModule collection
    # is passed; None for stock single-module runs, which fall back to the global
    # mpu groups below (byte-identical behavior).

File: megatron/training/training.py [hunk 6]
```diff
@@ -1186,7 +1194,6 @@ def pretrain(
     else:
         checkpointing_context = {}
 
-    # Model, optimizer, and learning rate.
```
Comment (yashaswikarnati) [inline megatron/training/training.py:1189]: why this comment is removed ?

File: megatron/training/training.py [hunk 7]
```diff
@@ -1076,8 +1080,12 @@ def pretrain(
...
         append_to_progress_log(args.save, "Starting job")
 
-    # Set pytorch JIT layer fusion options and warmup JIT functions.
-    set_jit_fusion_options()
+    _jit_tp_size = (
+        get_pg_size(schedule_pg_collection.get_language_model_collection().tp)
+        if schedule_pg_collection is not None and schedule_pg_collection.has_language_model()
+        else None
+    )
+    set_jit_fusion_options(tp_size=_jit_tp_size)
```
Comment (yashaswikarnati) [inline megatron/training/training.py:1097]: what happens to this on encoder ranks ?

File: megatron/training/checkpointing.py [hunk 8]
```diff
@@ -664,8 +687,9 @@ def save_checkpoint(iteration, model, optimizer, opt_param_scheduler, num_floati
 
                 if args.ckpt_fully_parallel_save:
                     if args.ckpt_fully_parallel_save_process_group == 'dp':
-                        process_group = mpu.get_data_parallel_group(with_context_parallel=True)
+                        process_group = dp_cp_group if dp_cp_group is not None else mpu.get_data_parallel_group(with_context_parallel=True)
                     elif args.ckpt_fully_parallel_save_process_group == 'ep_dp':
+                        # NOTE: expert-DP save group not threaded yet; falls back to mpu (not on the disjoint-grid path).
```
Comment (yashaswikarnati) [inline megatron/training/checkpointing.py]: why # NOTE: expert-DP save group not threaded yet; falls back to mpu (not on the disjoint-grid path).

 ?

File: megatron/training/checkpointing.py [hunk 9]
```diff
@@ -2033,7 +2064,12 @@ def load_model_state_dict(module, state_dict, strict: bool):
 
                 # access rng_state for data parallel rank
                 if args.data_parallel_random_init:
-                    rng_state = rng_state[mpu.get_data_parallel_rank()]
+                    dp_cp_rank = (
```
Comment (yashaswikarnati) [inline megatron/training/checkpointing.py]: this should be dp rank? not dp cp ?

File: megatron/training/initialize.py [hunk 10]
```diff
@@ -93,18 +95,24 @@ def state_restore_func(state_dict):
...
+            get_embedding_ranks,
+            get_position_embedding_ranks,
+            store,
+            skip_model_parallel_init=skip_model_parallel_init,
         )
 
+        # Random seeds (skipped when caller owns seeding -- reads mpu ranks).
+        if not skip_model_parallel_init:
+            print_rank_0("> setting random seeds to {} ...".format(args.seed))
+            _set_random_seed(
```
Comment (yashaswikarnati) [inline megatron/training/initialize.py]: we dont need to gaurd the set random seed in not skip model parallel init path ? we threaded process groups in both global process groups and also pass custom ones ?  why is this guarded here ?

File: megatron/training/checkpointing.py [hunk 11]
```diff
@@ -609,9 +629,9 @@ def save_checkpoint(iteration, model, optimizer, opt_param_scheduler, num_floati
     # the dense and expert parallelism layouts disagree (e.g. TP > EP*ETP); the union
     # does, with at most one rank per (tp_rank, ep_rank) inside any DP group.
     if not torch.distributed.is_initialized() \
+            or ckpt_type != CheckpointType.LEGACY \
             or mpu.get_data_parallel_rank() == 0 \
-            or mpu.get_expert_data_parallel_rank() == 0 \
-            or ckpt_type != CheckpointType.LEGACY:
+            or mpu.get_expert_data_parallel_rank() == 0:
```
Comment (yashaswikarnati) [inline megatron/training/checkpointing.py]: whats the change at this line? why are we still using mpu groups here ?

**Review by deepakn94 — APPROVED** [review]

**Review by JRD971000 — APPROVED** [review]
discussed over slack, LGTM!

**Review by Phlip79 — APPROVED** [review]

**Review by cuichenx — APPROVED** [review]

**Comment by yashaswikarnati** [comment]
/ok to test [1bdd7a9](https://github.com/NVIDIA/Megatron-LM/pull/5486/commits/1bdd7a9442f2eaaefb631aa3bfcf0a86446bd560)

**Comment by yashaswikarnati** [comment]
/ok to test  [2b23843](https://github.com/NVIDIA/Megatron-LM/pull/5486/commits/2b23843003d473256293ab766a05a156a88d26d2)

**Comment by yashaswikarnati** [comment]
/ok to test [2b23843](https://github.com/NVIDIA/Megatron-LM/pull/5486/commits/2b23843003d473256293ab766a05a156a88d26d2)

**Review by yaoyu-33 — APPROVED** [review]
